### PR TITLE
Buffer wrapping protection

### PR DESCRIPTION
**Pull request by doddi**

### CODE REVIEW

File: anathema-backend/src/tui/buffer.rs [hunk 1]
```diff
@@ -390,4 +410,75 @@ mod test {
...
+
+    #[test]
+    fn get_checks_range() {
+        let mut under_test = Buffer::new((1, 2));
+
+        under_test.put_glyph(Glyph::from_char('1', 1), LocalPos::new(0, 1));
+        under_test.put_glyph(Glyph::from_char('2', 1), LocalPos::new(0, 2));
+
+        let invalid_pos = LocalPos::new(1, 0);
+        assert_eq!(under_test.get(invalid_pos), None);
```
Comment (FishingHacks) [inline anathema-backend/src/tui/buffer.rs:455]:
```suggestion
        let mut under_test = Buffer::new((1, 2));

        assert_eq!(under_test.get(LocalPos::new(1, 0)), None);
        assert_eq!(under_test.get(LocalPos::new(0, 2)), None);
```

File: anathema-backend/src/tui/buffer.rs [hunk 2]
```diff
@@ -390,4 +410,75 @@ mod test {
...
+
+        assert_eq!(under_test.get(valid_pos).unwrap().1.clone(), Style::reset());
+    }
+
+    #[test]
+    fn put_glyph_checks_range() {
+        let mut under_test = Buffer::new((1, 2));
+
+        let valid_pos = LocalPos::new(0, 1);
+        under_test.put_glyph(Glyph::from_char('1', 1), LocalPos::new(0, 1));
```
Comment (FishingHacks) [inline anathema-backend/src/tui/buffer.rs:438]:
```suggestion
        under_test.put_glyph(Glyph::from_char('1', 1), valid_pos);
```

File: anathema-backend/src/tui/buffer.rs [hunk 3]
```diff
@@ -390,4 +410,75 @@ mod test {
...
+        under_test.put_glyph(Glyph::from_char('2', 1), LocalPos::new(0, 2));
+
+        let invalid_local_pos = LocalPos::new(1, 0);
+        under_test.put_glyph(Glyph::from_char('3', 1), invalid_local_pos);
```
Comment (FishingHacks) [inline anathema-backend/src/tui/buffer.rs:442]:
```suggestion
        under_test.put_glyph(Glyph::from_char('3', 1), LocalPos::new(1, 0));
```

File: anathema-backend/src/tui/buffer.rs [hunk 4]
```diff
@@ -390,4 +410,75 @@ mod test {
...
+        under_test.put_glyph(Glyph::from_char('2', 1), valid_pos);
+
+        let invalid_local_pos = LocalPos::new(1, 0);
+        let new_style = Style {
+            fg: Some(Color::Red),
+            bg: None,
+            attributes: tui::Attributes::empty(),
+        };
+        under_test.update_cell(new_style, invalid_local_pos);
+
```
Comment (FishingHacks) [inline anathema-backend/src/tui/buffer.rs:429]:
```suggestion

        let new_style = Style {
                fg: Some(Color::Blue),
                bg: None,
                attributes: tui::Attributes::empty(),
        };
        under_test.update_cell(new_style, LocalPos::new(0, 2))

```

File: anathema-backend/src/tui/buffer.rs [hunk 5]
```diff
@@ -390,4 +410,75 @@ mod test {
...
+        under_test.put_glyph(Glyph::from_char('1', 1), valid_pos);
+        under_test.put_glyph(Glyph::from_char('2', 1), valid_pos);
+
+        let invalid_local_pos = LocalPos::new(1, 0);
+        let new_style = Style {
+            fg: Some(Color::Red),
+            bg: None,
+            attributes: tui::Attributes::empty(),
+        };
+        under_test.update_cell(new_style, invalid_local_pos);
```
Comment (FishingHacks) [inline anathema-backend/src/tui/buffer.rs:428]:
```suggestion
        let new_style = Style {
            fg: Some(Color::Red),
            bg: None,
            attributes: tui::Attributes::empty(),
        };
        under_test.update_cell(new_style, LocalPos::new(1, 0));
```

**Review by FishingHacks — CHANGES_REQUESTED** [review]
Very good PR!

**Comment by FishingHacks** [comment]
Would you mind also fixing these issues for the `canvas` widget buffer? See: https://github.com/togglebyte/anathema/blob/dev/anathema-default-widgets/src/canvas.rs#L18

File: anathema-backend/src/tui/buffer.rs (hunk 5 above)
Comment (doddi) [reply anathema-backend/src/tui/buffer.rs]: https://github.com/togglebyte/anathema/pull/73/commits/ed34f4b31d3cb56651c6563c263d140196f9ab68

File: anathema-backend/src/tui/buffer.rs (hunk 4 above)
Comment (doddi) [reply anathema-backend/src/tui/buffer.rs:427]: https://github.com/togglebyte/anathema/pull/73/commits/ed34f4b31d3cb56651c6563c263d140196f9ab68

File: anathema-backend/src/tui/buffer.rs (hunk 3 above)
Comment (doddi) [reply anathema-backend/src/tui/buffer.rs]: https://github.com/togglebyte/anathema/pull/73/commits/ed34f4b31d3cb56651c6563c263d140196f9ab68

File: anathema-backend/src/tui/buffer.rs (hunk 2 above)
Comment (doddi) [reply anathema-backend/src/tui/buffer.rs:435]: https://github.com/togglebyte/anathema/pull/73/commits/ed34f4b31d3cb56651c6563c263d140196f9ab68

File: anathema-backend/src/tui/buffer.rs (hunk 1 above)
Comment (doddi) [reply anathema-backend/src/tui/buffer.rs]: https://github.com/togglebyte/anathema/pull/73/commits/ed34f4b31d3cb56651c6563c263d140196f9ab68

**Review by FishingHacks — APPROVED** [review]
looks good!

**Comment by doddi** [comment]
The changes made are to the underlying buffer and not the canvas. There is a small performance issue whereby the canvas is marked dirty which would cause an unnecessary paint(?) when out of range updates are made.
Im not sure if that should be addressed/needed as it is already getting used 'incorrectly'

**Pull request merged**